src/Features/Core/Portable/PdbSourceDocument/AbstractPdbSourceDocumentNavigationService.cs

```diff
@@ -0,0 +1,256 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.PdbSourceDocument
+{
+    internal abstract class AbstractPdbSourceDocumentNavigationService : IPdbSourceDocumentNavigationService
+    {
+        private MetadataAsSourceWorkspace? _workspace;
+        private readonly IPdbFileLocatorService _pdbFileLocatorService;
+        private readonly IPdbSourceDocumentLoaderService _pdbSourceDocumentLoaderService;
+
+        protected abstract string LanguageName { get; }
+
+        protected AbstractPdbSourceDocumentNavigationService(IPdbFileLocatorService pdbFileLocatorService, IPdbSourceDocumentLoaderService pdbSourceDocumentLoaderService)
+        {
+            _pdbFileLocatorService = pdbFileLocatorService;
+            _pdbSourceDocumentLoaderService = pdbSourceDocumentLoaderService;
+        }
+
+        public async Task<MetadataAsSourceFile?> GetPdbSourceDocumentAsync(Project project, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            var peReference = compilation.GetMetadataReference(symbol.ContainingAssembly) as PortableExecutableReference;
+            if (peReference is null)
+                return null;
+
+            var dllPath = peReference.FilePath;
+            if (dllPath is null)
+                return null;
+
+            using var pdbStream = await _pdbFileLocatorService.GetPdbPathAsync(dllPath, cancellationToken).ConfigureAwait(false);
+            if (pdbStream is null)
+                return null;
+
+            using var dllStream = File.OpenRead(dllPath);
+            var filePaths = GetSourcePaths(symbol, dllStream, pdbStream);
+            if (filePaths.Length == 0)
+                return null;
+
+            // TODO: Do we need our own workspace?
+            //       We should probably cache things from the same document
+            //       Does each assembly get its own project added to that workspace?
+            if (_workspace == null)
+            {
+                _workspace = new MetadataAsSourceWorkspace(null!, project.Solution.Workspace.Services.HostServices);
+            }
+
+            var symbolId = SymbolKey.Create(symbol, cancellationToken);
+            var projectId = ProjectId.CreateNewId();
+            var languageServices = _workspace.Services.GetLanguageServices(LanguageName);
+            var compilationOptions = languageServices.CompilationFactory!.GetDefaultCompilationOptions().WithOutputKind(OutputKind.DynamicallyLinkedLibrary);
+            var parseOptions = languageServices.GetRequiredService<ISyntaxTreeFactoryService>().GetDefaultParseOptionsWithLatestLanguageVersion();
```

Comment:
We need to extract the options from the binary.  See [https://github.com/dotnet/roslyn/blob/main/docs/features/pdb-compilation-options.md](https://github.com/dotnet/roslyn/blob/main/docs/features/pdb-compilation-options.md)